src/librustc_codegen_llvm/context.rs

```diff
@@ -156,6 +172,13 @@ pub unsafe fn create_module(
     if llvm_util::get_major_version() < 9 {
         target_data_layout = strip_function_ptr_alignment(target_data_layout);
     }
+    if sess.target.target.arch == "x86" || sess.target.target.arch == "x86_64" {
```

Comment:
We should do either one or the other, not both, if either one at all. i.e. we ought to pick whether our target data layouts strings are llvm10 (which IMO they should be) or something else. Then we don’t need code that can both "upgrade" and "downgrade" the string.

Reply:
There's three parts here:

 * Data layout in our own target definitions: I agree that we should update those to use the LLVM 10 data layouts.
 * Downgrade support: If we use LLVM 10 data layout, then we must have downgrade support, otherwise older LLVM versions cannot be used anymore.
 * Upgrade support: Upgrade support is needed if we want to keep 3rd party target definitions (for x86 based targets) working. If we're okay with breaking these and have no stability guarantees in this area, then we can drop the upgrade support.

Reply:
I don’t believe we have any stability guarantees for external target files. Going in that direction would necessitate removing at least some fields from target specs and data layout would be one of those fields -- we dont want to be on the hook for ensuring stability of LLVM's interfaces.

With that in mind I'd be happy with this if we only kept the downgrade code.

Reply:
I've dropped the upgrade code and updated all the in-tree targets ... hopefully without typos ;)